### PR TITLE
refactor(pipelines transform): use array instead of map for pipelines

### PR DESCRIPTION
**Pull request by jdrouet**
Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

As requested in this issue https://github.com/vectordotdev/vector/issues/11921, the goal of this PR is to remove the map in the pipeline configuration to reduce verbosity.
This removes:
- the `order` field in the configuration
- the ability to put every single pipeline in a different file.

As discussed in the issue, this breaks backward compatibility and it's fine.

### CODE REVIEW

**Comment by jdrouet** [comment]
The soak tests being broken is a normal thing: this change breaks backward compatibility and therefore the current configuration files can't be used with the version on `master`.

File: src/transforms/pipelines/config.rs [hunk 1]
```diff
@@ -169,15 +145,11 @@ impl EventTypeConfig {
...
-            let pipeline_name = name.join(&pipe_name);
-            let pipeline_config = self
-                .pipelines
-                .get_mut(&pipe_name)
-                .ok_or_else(|| format!("Unable to find pipeline with name {:?}", pipe_name))?;
+        for (pipeline_index, pipeline_config) in self.0.iter_mut().enumerate() {
+            let pipeline_name = name.join(pipeline_index);
             let topology = pipeline_config
                 .expand(&pipeline_name, &next_inputs)?
-                .ok_or_else(|| format!("Unable to expand pipeline named {:?}", pipe_name))?;
```
Comment (jszwedko) [inline src/transforms/pipelines/config.rs:180]: I think using the pipeline name field here might make it easier for users to find.

File: src/transforms/pipelines/config.rs [hunk 2]
```diff
@@ -103,35 +103,27 @@ impl PipelineConfig {
...
-        for (name, pipeline) in self.pipelines.iter() {
-            for (index, transform) in pipeline.transforms.iter().enumerate() {
+        for (pipeline_index, pipeline) in self.0.iter().enumerate() {
+            for (transform_index, transform) in pipeline.transforms.iter().enumerate() {
                 if !transform.nestable(parents) {
                     return Err(format!(
-                        "the transform {} in pipeline {:?} cannot be nested in {:?}",
-                        index, name, parents
+                        "the transform {} in pipeline {} cannot be nested in {:?}",
+                        transform_index, pipeline_index, parents
```
Comment (jszwedko) [inline src/transforms/pipelines/config.rs:126]: Could we maybe pull the `name` field of the pipeline and the `id` of the transform? I think that would make it easier for users to find the offending transform.

**Review by jszwedko — COMMENTED** [review]
Looks good! Just a couple of comments about error messages.

**Comment by jdrouet** [comment]
@jszwedko I updated the error messages to have both the index and the pipeline name. Tell my if you were something of something else 😉

**Review by binarylogic — APPROVED** [review]

**Review by blt — APPROVED** [review]
This strikes me as straightforward and tidier in the config.

**Review by jszwedko — APPROVED** [review]
Looks good, thanks @jdrouet !

**Comment by jszwedko** [comment]
Merging since soak tests won't run due to configuration change being incompatible with master.

**Comment by jszwedko** [comment]
@jdrouet stumbled upon this comment which needs updating from these changes: https://github.com/vectordotdev/vector/blob/1052e95bb1534e606d8756bcf138df1261687de0/src/transforms/pipelines/mod.rs#L6

**Pull request merged**